Data relay server

pingpingxiu-DOT-ca-gov opened the pull request:
This is the initial check in of the data relay services

Currently due to the extensive refactoring for this PR, the service is not completely connected, 
and at this moment (mar 8) this is intended for early feedback. 

Reading order of the PR: 
 Go to the location of the folder (data-relay)
and first read the README.md, and understand the diagram. 
Then, navigate to the individual files.

pingpingxiu-DOT-ca-gov commented:
Thanks @ian-r-rose  for introducing to me the `mermaidjs`. 
I adopted it in this PR, seems pretty handy on expressing ideas using visualized diagrams.

ian-r-rose commented on jobs/data_relay/table_aws_uploader.py:
```diff
@@ -0,0 +1,538 @@
+        ssh_command = (
+            f"/usr/local/bin/aws s3 cp {parquet_output} "
+            "s3://caltrans-pems-dev-us-west-2-raw/db96_export_staging_area"
+            f"/tables/{final_topic}/{date_folder_substring}{district_folder_substring}"
```
Are these parquet files batched in any way so that these files are a single day's worth of data per district? I don't see much in the way of batching, so I'd expect subsequent files to overwrite the older ones.

ian-r-rose reviewed (commented):
Thanks @pingpingxiu-DOT-ca-gov, this is a tremendous amount of work. I'm still digesting, but wanted to give you some early thoughts.

A few high level questions:

1. Have you tested this pipeline end-to-end? I don't see any files from this in the S3 bucket.
1. I'm not very familiar with Kafka myself, but there are a large number of moving pieces here, which don't necessarily seem worth the added complexity to me:
    * Kafka
    * Confluent (?)
    * Zookeeper
    * Kafka schema registry
    * Elasticsearch
1. Is anything batching the files? How many do we expect to get uploaded to S3 every day?

ian-r-rose commented:
> Thanks @ian-r-rose for introducing to me the `mermaidjs`. I adopted it in this PR, seems pretty handy on expressing ideas using visualized diagrams.

Glad you found it helpful!

pingpingxiu-DOT-ca-gov commented:
```
A few high level questions:

Have you tested this pipeline end-to-end? I don't see any files from this in the S3 bucket.
```
The system before refactoring, did produce and upload some data to that AWS bucket, but on different set of tables (DataWarehourse rather than DB96) and with json rather than parquet. Since old format no longer valid per our discussion, I turned it off a few days ago. 
 
The new system, I'm still working on plumbing end-2-end. However, I thought it would give our PR process more time for valued conversations, so I actually rushed to PR. For end-2-end, I hope I can get it on either today or tomorrow, and I'll let you know the folder it upload to. 

```
I'm not very familiar with Kafka myself, but there are a large number of moving pieces here, which don't necessarily seem worth the added complexity to me:
Kafka
Confluent (?)
Zookeeper
Kafka schema registry
Elasticsearch
```
Those components arguably simplifies the overall system & support workflow on a long term run. Kafka (by confluent) improves the data quality when comes to robust data transfer. Zookeeper ensures we have a clean environment
when comes to shared resources like data base connections. Schema Registry just offers a strongly-typed experience
that gives the rigor and quality. Elastic Search eliminates the local files. 

```
Is anything batching the files? How many do we expect to get uploaded to S3 every day?
```

As soon as enough data being crawled, the system will get upload to S3. 
The crawl-queue is generated on scheduled time, like daily or every 4 hours

I'm trying to do other responses when I have free time (have to catch up on another high priority project today)

pingpingxiu-DOT-ca-gov commented:
Hi @ian-r-rose , 
The latest commit is uploaded `8b04a33`
And one uploaded file for verify on your side: 

s3://caltrans-pems-dev-us-west-2-raw/db96_export_staging_area/tables/VDS30SEC/district=D10/year=202
4/month=3/day=13/D10.VDS30SEC_dump_20240313T233626Z_6589e633c78f5fa7.parquet

Adjusted the order of subfolders, increased the window size, fixed the casting issue, and padded the columns. 

The only thing left open is int parsing for LOOPn_SPD columns, because it is hard to 
get along with the NaN...

ian-r-rose commented:
The new file is looking close @pingpingxiu-DOT-ca-gov! I can confirm that there are 60,001 rows in the parquet file you uploaded. One thing I noticed is that the samples seem to be from 2024-03-11, but the file name indicates that they should be from 2024-03-13. Is that intentional?

pingpingxiu-DOT-ca-gov commented:
Yes, exactly `60,001`. Thanks for confirming that. 
I noticed that, this parquet file is less than 1 mega bytes, 1/10 of the original json size... 
If this aws cost a lot on storage, then 9/10 is saved by your tip:)


And the timestamp on the file naming, that's negotiable. 

Question for you, do we have the adhoc use case on specifying more specific subpath (including the district, year, month or even day)? If so, then misaligned timestamps would cause an issue because we don't get what we specified. 

On the other hand, my impression is, since in snowflake/dbt environment, we basically consumes the data from the staging, 
rather than working on plumbing work into the staging. In another word, we sits at the downstream of the staging. We would use sql query everywhere on snowflake, using table names, rather than file paths. 

If we don't use file paths, does the naming matter?  

@ian-r-rose

pingpingxiu-DOT-ca-gov commented:
Ok, I'll make the change. Anticipate a new commit later in the day.

pingpingxiu-DOT-ca-gov commented:
New commit `27ecf37b` submitted, 
"using logic date in parquet physical file name"

And one latest uploads to verify: 

`s3://caltrans-pems-dev-us-west-2-raw/db96_export_staging_area/tables/VDS30SEC/district=D4/year=2024/month=3/day=13/D4.VDS30SEC_dump_20240314T205035Z_b7009b83de08cc50.parquet`

Let me know how to proceed, @ian-r-rose

pingpingxiu-DOT-ca-gov commented:
By the way, the latest commit `2a8f3d0` is on configuration of the automated jobs. 
Only the append_to_crawl_queue is scheduled every four hour, 
and the other two components (oracle_crawl_daemon and aws_uploader) are always-running, 
hence, the cron job is merely for checking whether they're down and re-launch them.

ian-r-rose commented:
Thanks @pingpingxiu-DOT-ca-gov. The most recent parquet file you uploaded seems to still have some issues with the timestamp (I'm seeing dates from 1970):

<img width="1170" alt="image" src="https://github.com/cagov/caldata-mdsa-caltrans-pems/assets/5728311/1e195274-f999-4cb6-874b-2ca947b1ffda">


Is that the file I should be looking at? The previous one seemingly had correct timestamps

pingpingxiu-DOT-ca-gov commented:
This is the latest `s3://caltrans-pems-dev-us-west-2-raw/db96_export_staging_area/tables/VDS30SEC/district=D4/year=2024/month=3/day=13/D4.VDS30SEC_dump_20240315T210643Z_e01e58d26a44590d.parquet`

pingpingxiu-DOT-ca-gov commented:
@ian-r-rose

ian-r-rose commented:
The new one is looking pretty good! The only thing that I see that could improve the schema would be to convert the volume columns to nullable ints rather than doubles.

pingpingxiu-DOT-ca-gov commented:
`s3://caltrans-pems-dev-us-west-2-raw/db96_export_staging_area/tables/VDS30SEC/district=D4/year=2024/month=3/day=13/D4.VDS30SEC_dump_20240315T225626Z_1cdf5049d550258f.parquet`

ian-r-rose commented:
Hi @pingpingxiu-DOT-ca-gov, any updates here? I'm still not seeing any data since March 14

pingpingxiu-DOT-ca-gov commented:
Starting `2024-03-20T14-0700`, these were the entire uploading events, logged in my service log. 



```
2024-03-20T14:49:25.570979-0700 INFO table_aws_uploader.py:1062 Executing command/usr/local/bin/aws s3 cp /nfsdata/dataop/uploader/tmp/
D4.VDS30SEC/D4.VDS30SEC_dump_20240320T214904Z_654c8fd9f221e978.parquet s3://caltrans-pems-dev-us-west-2-raw/db96_export_staging_area/ta
bles/VDS30SEC/district=D4/year=2024/month=3/day=14/ --ca-bundle /etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt
2024-03-20T14:49:45.238329-0700 INFO table_aws_uploader.py:1062 Executing command/usr/local/bin/aws s3 cp /nfsdata/dataop/uploader/tmp/
D4.VDS30SEC/D4.VDS30SEC_dump_20240320T214926Z_ac61451defa0b339.parquet s3://caltrans-pems-dev-us-west-2-raw/db96_export_staging_area/ta
bles/VDS30SEC/district=D4/year=2024/month=3/day=14/ --ca-bundle /etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt
2024-03-20T14:50:05.170439-0700 INFO table_aws_uploader.py:1062 Executing command/usr/local/bin/aws s3 cp /nfsdata/dataop/uploader/tmp/
D4.VDS30SEC/D4.VDS30SEC_dump_20240320T214946Z_f660c28a74585b5b.parquet s3://caltrans-pems-dev-us-west-2-raw/db96_export_staging_area/ta
bles/VDS30SEC/district=D4/year=2024/month=3/day=14/ --ca-bundle /etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt
2024-03-20T14:50:25.092416-0700 INFO table_aws_uploader.py:1062 Executing command/usr/local/bin/aws s3 cp /nfsdata/dataop/uploader/tmp/
D4.VDS30SEC/D4.VDS30SEC_dump_20240320T215006Z_59257b66c65a78b7.parquet s3://caltrans-pems-dev-us-west-2-raw/db96_export_staging_area/ta
bles/VDS30SEC/district=D4/year=2024/month=3/day=14/ --ca-bundle /etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt
2024-03-20T14:50:45.219976-0700 INFO table_aws_uploader.py:1062 Executing command/usr/local/bin/aws s3 cp /nfsdata/dataop/uploader/tmp/
D4.VDS30SEC/D4.VDS30SEC_dump_20240320T215026Z_01e2677fc0967c48.parquet s3://caltrans-pems-dev-us-west-2-raw/db96_export_staging_area/ta
bles/VDS30SEC/district=D4/year=2024/month=3/day=14/ --ca-bundle /etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt
2024-03-20T16:48:17.486970-0700 INFO table_aws_uploader.py:1062 Executing command/usr/local/bin/aws s3 cp /nfsdata/dataop/uploader/tmp/
D3.VDS30SEC/D3.VDS30SEC_dump_20240320T234804Z_a9cf3e2438922a74.parquet s3://caltrans-pems-dev-us-west-2-raw/db96_export_staging_area/ta
bles/VDS30SEC/district=D3/year=2024/month=3/day=14/ --ca-bundle /etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt
2024-03-20T17:33:59.874139-0700 INFO table_aws_uploader.py:1062 Executing command/usr/local/bin/aws s3 cp /nfsdata/dataop/uploader/tmp/
D4.VDS30SEC/D4.VDS30SEC_dump_20240320T234819Z_21b93a5fbcdfea92.parquet s3://caltrans-pems-dev-us-west-2-raw/db96_export_staging_area/ta
bles/VDS30SEC/district=D4/year=2024/month=3/day=14/ --ca-bundle /etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt
2024-03-20T18:23:22.521958-0700 INFO table_aws_uploader.py:1062 Executing command/usr/local/bin/aws s3 cp /nfsdata/dataop/uploader/tmp/
D3.VDS30SEC/D3.VDS30SEC_dump_20240321T010002Z_48d457f1c62052d0.parquet s3://caltrans-pems-dev-us-west-2-raw/db96_export_staging_area/ta
bles/VDS30SEC/district=D3/year=2024/month=3/day=14/ --ca-bundle /etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt
2024-03-20T19:17:16.967450-0700 INFO table_aws_uploader.py:1062 Executing command/usr/local/bin/aws s3 cp /nfsdata/dataop/uploader/tmp/
D4.VDS30SEC/D4.VDS30SEC_dump_20240321T012326Z_7a77ea8ebff760f3.parquet s3://caltrans-pems-dev-us-west-2-raw/db96_export_staging_area/ta
bles/VDS30SEC/district=D4/year=2024/month=3/day=14/ --ca-bundle /etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt
2024-03-20T19:33:05.246926-0700 INFO table_aws_uploader.py:1062 Executing command/usr/local/bin/aws s3 cp /nfsdata/dataop/uploader/tmp/
D5.VDS30SEC/D5.VDS30SEC_dump_20240321T021722Z_eb6ef6af95fac7b9.parquet s3://caltrans-pems-dev-us-west-2-raw/db96_export_staging_area/ta
bles/VDS30SEC/district=D5/year=2024/month=3/day=14/ --ca-bundle /etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt
2024-03-20T19:50:02.629887-0700 INFO table_aws_uploader.py:1062 Executing command/usr/local/bin/aws s3 cp /nfsdata/dataop/uploader/tmp/
D6.VDS30SEC/D6.VDS30SEC_dump_20240321T023308Z_91bae7deadee863f.parquet s3://caltrans-pems-dev-us-west-2-raw/db96_export_staging_area/ta
bles/VDS30SEC/district=D6/year=2024/month=3/day=14/ --ca-bundle /etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt
2024-03-21T01:16:55.585648-0700 INFO table_aws_uploader.py:1062 Executing command/usr/local/bin/aws s3 cp /nfsdata/dataop/uploader/tmp/
D11.VDS30SEC/D11.VDS30SEC_dump_20240321T072048Z_f707ac9e0faf1359.parquet s3://caltrans-pems-dev-us-west-2-raw/db96_export_staging_area/
tables/VDS30SEC/district=D11/year=2024/month=3/day=19/ --ca-bundle /etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt
2024-03-21T02:13:01.390292-0700 INFO table_aws_uploader.py:1062 Executing command/usr/local/bin/aws s3 cp /nfsdata/dataop/uploader/tmp/
D12.VDS30SEC/D12.VDS30SEC_dump_20240321T081701Z_bff2aea651f44efc.parquet s3://caltrans-pems-dev-us-west-2-raw/db96_export_staging_area/
tables/VDS30SEC/district=D12/year=2024/month=3/day=19/ --ca-bundle /etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt
2024-03-21T09:56:25.494062-0700 INFO table_aws_uploader.py:1062 Executing command/usr/local/bin/aws s3 cp /nfsdata/dataop/uploader/tmp/
D5.VDS30SEC/D5.VDS30SEC_dump_20240321T160546Z_6b783ba3d41d1c2c.parquet s3://caltrans-pems-dev-us-west-2-raw/db96_export_staging_area/ta
bles/VDS30SEC/district=D5/year=2024/month=3/day=19/ --ca-bundle /etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt
2024-03-21T11:03:18.854635-0700 INFO table_aws_uploader.py:1062 Executing command/usr/local/bin/aws s3 cp /nfsdata/dataop/uploader/tmp/
D6.VDS30SEC/D6.VDS30SEC_dump_20240321T165630Z_efc4155d5aa50e27.parquet s3://caltrans-pems-dev-us-west-2-raw/db96_export_staging_area/ta
bles/VDS30SEC/district=D6/year=2024/month=3/day=19/ --ca-bundle /etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt
```

ian-r-rose commented:
I see those files, but I expected more. If we're running this every six hours, I expect 4*9 = 36 files per day. I only see four files from the 19th, and then none thereafter, per your log.

pingpingxiu-DOT-ca-gov commented:
For initial plumbing I set the batch for 12 hours, just to test the limits. Once I get some understanding of the capacity, 
I'll dial back to the 6 hours. This process may take a few days to stablize. 

It won't affect the consuming of those data on your side.

ian-r-rose commented:
Okay, thanks @pingpingxiu-DOT-ca-gov. Let me know when it's stabilized a bit, and I can get to work on loading the data to Snowflake

ian-r-rose commented:
@pingpingxiu-DOT-ca-gov any updates here? I see that some files are being uploaded to S3, but they are not quite I was expecting. Taking a look at the data for Monday in District 8 as an example (`db96_export_staging_area/tables/VDS30SEC/district=D8/year=2024/month=3/day=25`):

<img width="845" alt="image" src="https://github.com/cagov/caldata-mdsa-caltrans-pems/assets/5728311/93033e27-aa62-4997-b936-a5a0247f2f2f">

1. I see two files in this folder. I expected to see ~4 files (if we're trying to do ~6 hour batches).  This seems to be the case for most of the districts and dates. What batching strategy are you using right now?
2. Based on the timestamps and last modified dates in the files, it looks like they are being uploaded about 24 hours *after* the logical date. Is that what you expect? It's not necessarily a problem, but I was under the impression we'd be uploading the files sooner than that.
3. Taking a look at those two files, the data in them don't cover the whole day. The first file has values from 02:12:19 to 11:59:57, and the second file has values from 11:59:57 to 19:59:47. This leaves approximately six hours with no values. Do you have any ideas what might be going on there?

pingpingxiu-DOT-ca-gov commented:
I was doing extensive troubleshooting now. 
There were undeterministic out of memory error coming with large batch data exceeding 2 million records, from pandas data frame operation, and I have tried different ways of mitigating this. 

Right now I'm limiting the cap of the batch to be 1000000 records, which I haven't observed causing OOM errors so far. 
And I need to figure out a way to confirm the uploading success for a batch to allow for retrying, in case a batch failed to upload. 

@ian-r-rose

ian-r-rose commented:
Hi @pingpingxiu-DOT-ca-gov I have a few questions about the timestamps for the files you are uploading (example from d12 shown here):

<img width="1179" alt="image" src="https://github.com/cagov/caldata-mdsa-caltrans-pems/assets/5728311/f044a1b6-386b-454f-8fda-120175bcbb5f">

1. Based on the file upload times, it looks like the files are being uploaded about 24 hours _after_ the times within the files. Is this what you expect?
2. One file has samples for about 2 hours from ~00:00 to ~02:00. The next has samples from ~02:00 to ~08:00. This isn't really evenly balanced. Do you know why that is?
3. What time zone are these samples in? In the file they are TZ-naive (no time zone). Do we expect PST? PDT? UTC? Getting this right will be extremely important for interpretability.

ian-r-rose commented:
@pingpingxiu-DOT-ca-gov any updates here? I see that data has stopped uploading since about Friday. I also still have questions about the timestamps in the data, can you look into those?

pingpingxiu-DOT-ca-gov commented:
I will get all those addressed by tomorrow morning.

From: Ian Rose ***@***.***>
Sent: Tuesday, April 9, 2024 2:23 PM
To: cagov/caldata-mdsa-caltrans-pems ***@***.***>
Cc: Xiu, ***@***.*** ***@***.***>; Mention ***@***.***>
Subject: Re: [cagov/caldata-mdsa-caltrans-pems] Data relay server (PR #105)

EXTERNAL EMAIL. Links/attachments may not be safe.

@pingpingxiu-DOT-ca-gov<https://urldefense.com/v3/__https:/github.com/pingpingxiu-DOT-ca-gov__;!!LWi6xHDyrA!7lxwSyqyoaIFZZvMnXIZrzn9HYQAw8Ix3Xm_1BfBybORKHFm0sk0XdFuAykOMHwU9eEitYjhtdiAh3vYW_2_M4ly18LeTFo$> any updates here? I see that data has stopped uploading since about Friday. I also still have questions about the timestamps in the data, can you look into those?

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https:/github.com/cagov/caldata-mdsa-caltrans-pems/pull/105*issuecomment-2046073041__;Iw!!LWi6xHDyrA!7lxwSyqyoaIFZZvMnXIZrzn9HYQAw8Ix3Xm_1BfBybORKHFm0sk0XdFuAykOMHwU9eEitYjhtdiAh3vYW_2_M4lyhCLhz_k$>, or unsubscribe<https://urldefense.com/v3/__https:/github.com/notifications/unsubscribe-auth/BDVOYUKN7UXK6ASAMW7ST33Y4RLZ7AVCNFSM6AAAAABENVZ5IKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANBWGA3TGMBUGE__;!!LWi6xHDyrA!7lxwSyqyoaIFZZvMnXIZrzn9HYQAw8Ix3Xm_1BfBybORKHFm0sk0XdFuAykOMHwU9eEitYjhtdiAh3vYW_2_M4lyE701Knk$>.
You are receiving this because you were mentioned.Message ID: ***@***.******@***.***>>

ian-r-rose commented:
Thanks!

pingpingxiu-DOT-ca-gov commented:
@ian-r-rose 
The Oracle Database seems to be using the pure local time. It is even taking care of PST to PDT by itself. 

Here is the reasoning process:  

We design three test queries: (1), (2), (3)

(1)
SELECT * FROM ( SELECT * FROM D3.VDS30SEC WHERE SAMPLE_TIME BETWEEN TO_DATE('**2024-03-10 01:30:00**', 'YYYY-MM-DD HH24:MI:SS') AND TO_DATE('**2024-03-10 01:40:00**', 'YYYY-MM-\
DD HH24:MI:SS') ORDER BY SAMPLE_TIME DESC ) WHERE rownum < 10
Has Results

(2)
SELECT * FROM ( SELECT * FROM D3.VDS30SEC WHERE SAMPLE_TIME BETWEEN TO_DATE('**2024-03-10 02:30:00**', 'YYYY-MM-DD HH24:MI:SS') AND TO_DATE('**2024-03-10 02:40:00**', 'YYYY-MM-\
DD HH24:MI:SS') ORDER BY SAMPLE_TIME DESC ) WHERE rownum < 10
Has **No Results**

(3)
SELECT * FROM ( SELECT * FROM D3.VDS30SEC WHERE SAMPLE_TIME BETWEEN TO_DATE('**2024-03-10 03:30:00**', 'YYYY-MM-DD HH24:MI:SS') AND TO_DATE('**2024-03-10 03:40:00**', 'YYYY-MM-\
DD HH24:MI:SS') ORDER BY SAMPLE_TIME DESC ) WHERE rownum < 10
Has Results

Assuming the Oracle stick to *PST only* or *PDT only*, then the query (2) wouldn't have been returning None. 
The fact that 1 non-empty , 2 empty, 3 non-empty, can only happen when PST switches to PDT, and the 2AM-3AM 
is missing on that day (March 10th, 2024, the second Sunday of March).

pingpingxiu-DOT-ca-gov commented:
@ian-r-rose 
@ZhenyuZhu-Caltrans 

Today, add auto-git-issue creation for alarm, 
    add observability dashboard to track the relay status, 
    and troubleshooted the datetime issue in the script.

pingpingxiu-DOT-ca-gov commented:
Talked with @kengodleskidot , 
the TMP source data is `naive` -- no time zone indicator. 

So, all date time in the system should interpret date time as `naive`.

pingpingxiu-DOT-ca-gov commented:
The `table_aws_uploader.py` were ruff-ed, and the errors decreased from 150+ down to 30+. 
The new version was put into the last night's production test and verified. 

Remaining errors were not low-hanging fruits. Also, for other scripts like `oracle_puller_daemon.py` I hold off the ruff-ing to reduce the production test's impact scope. 

I suggest we ignore the rest of the formatting errors for this Pull Request. 

@ian-r-rose 
@ZhenyuZhu-Caltrans

ian-r-rose commented:
@pingpingxiu-DOT-ca-gov sounds good, I'll push a commit to this branch to disable ruff for your scripts.

ian-r-rose commented:
Okay @pingpingxiu-DOT-ca-gov I've turned off the linter for your data relay scripts, but left the formatter enabled. Does that work for you?

pingpingxiu-DOT-ca-gov commented:
thanks @ian-r-rose , 
I see all checks passed. 
As soon as @ZhenyuZhu-Caltrans  approve, I can merge
